### PR TITLE
feat: Supports the allowed viewer list for the canister log.

### PR DESCRIPTION
**Pull request by vincent-dfinity**
# Description

1. Supports the allowed viewer list for the canister log.
    Valid settings are:
     - `--add-log-viewer`, `--remove-log-viewer` and `--set-log-viewer` flags with `dfx canister update-settings` 
     - `--log-viewer` flag with `dfx canister create`
     - `canisters[].initialization_values.log_visibility.allowed_viewers` in `dfx.json`
2. Bump the replica to revision 285a5db07da50a4e350ec43bf3b488cc6fe36102 which includes the flag in ic-starter.

Fixes # (issue)

[SDK-1798](https://dfinity.atlassian.net/browse/SDK-1798)

# How Has This Been Tested?

- Added the e2e tests

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

[SDK-1798]: https://dfinity.atlassian.net/browse/SDK-1798?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/dfx-core/src/config/model/dfinity.rs
Comment (sesi200) [inline src/dfx-core/src/config/model/dfinity.rs:411]:
```suggestion
#[serde(rename_all = "snake_case")]
```
Yes, the rest of the file uses `"lowercase"`, but these are all one word fields. We use `snake_case` where it makes a difference

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (sesi200) [inline src/dfx/src/lib/canister_logs/log_visibility.rs:83]: I know it says todo. Hint: use `slog::warn!`

Comment (vincent-dfinity) [reply src/dfx/src/lib/canister_logs/log_visibility.rs]: Addressed at https://github.com/dfinity/sdk/pull/3921/commits/5a0deb608fc365a7d4c0628ff2d6478270d5322b

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (vincent-dfinity) [reply src/dfx/src/lib/canister_logs/log_visibility.rs]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/5a0deb608fc365a7d4c0628ff2d6478270d5322b.

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (maksymar) [inline src/dfx/src/lib/canister_logs/log_visibility.rs:24]: Doc text, the field name and the field type do not match. Doc text and field name suppose it only adds a single principal, while the type expects a vector of strings (which I assume is a vector of principals).

I'd expect it either to be plural or to be an optional string. Let me know if my assumptions are incorrect.

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (maksymar) [inline src/dfx/src/lib/canister_logs/log_visibility.rs:28]: ditto singular vs vec

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (maksymar) [inline src/dfx/src/lib/canister_logs/log_visibility.rs:38]: ditto singular vs vec

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (vincent-dfinity) [reply src/dfx/src/lib/canister_logs/log_visibility.rs:24]: You can add multiple viewers in one command like below

Comment (vincent-dfinity) [reply src/dfx/src/lib/canister_logs/log_visibility.rs:24]: This is why the action is `ArgAction::Append`.

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (vincent-dfinity) [reply src/dfx/src/lib/canister_logs/log_visibility.rs:12]: So here `log_visibility` is used to set to `public` and `controller`. So the usage is `--log-visibility controllers` and `--log-visibility public`. 

The intention is to have it unchanged to keep the backward compatibility.

File: CHANGELOG.md
Comment (maksymar) [inline CHANGELOG.md:11]: NIT: rephrasing to avoid duplicated which-which

`Added support for the canister log allowed viewer list, enabling specified users to access a canister's logs without needing to be set as the canister's controller.`

File: docs/cli-reference/dfx-canister.mdx
Comment (maksymar) [inline docs/cli-reference/dfx-canister.mdx:284]: > Can be either "controllers" or "public".

This does not seem to be the case any more, since it can be `allowed_viewers`. And it's not clear that `--log-viewer` modifies `log_visibility`.

File: e2e/tests-dfx/canister_logs.bash
Comment (maksymar) [inline e2e/tests-dfx/canister_logs.bash:82]: Maybe it makes sense to create Alice identity first, without adding it to allowed viewers, verify it has no access to logs, then add it as a viewer and verify it now has access.

File: e2e/tests-dfx/create.bash
Comment (maksymar) [inline e2e/tests-dfx/create.bash:356]: same 2 scenarios: users have no access & do have access.

File: e2e/tests-dfx/create.bash
Comment (maksymar) [inline e2e/tests-dfx/create.bash:386]: same 2 scenarios: users have no access & do have access.

File: docs/cli-reference/dfx-canister.mdx
Comment (maksymar) [inline docs/cli-reference/dfx-canister.mdx:1145]: Unlike `--set-controller` a new option `--set-log-viewer` also changes the setting of log visibility (if it was `public` or `controllers`). This message has to clarify it.

File: docs/cli-reference/dfx-canister.mdx
Comment (maksymar) [inline docs/cli-reference/dfx-canister.mdx:1141]: same comment as for `--set-log-viewer`

File: docs/cli-reference/dfx-canister.mdx
Comment (vincent-dfinity) [reply docs/cli-reference/dfx-canister.mdx:284]: I'd use `Specifies the principal as an allowed viewer. Can be specified more than once. Cannot be used with `--log-visibility`.  
As `a list of principals` might make users think they only need this parameter `once` to specify multiple principals. It's also consistent with the document for `--controller`.

File: e2e/tests-dfx/create.bash
Comment (vincent-dfinity) [reply e2e/tests-dfx/create.bash:356]: This test is used to cover the canister can be created with the allowed view list in dfx.json. The [create with default settings](https://github.com/dfinity/sdk/blob/a685c3f4f41a92da469ed0fa5a8e93d470c05e7b/e2e/tests-dfx/create.bash#L344) covers the no access case.

File: CHANGELOG.md
Comment (vincent-dfinity) [reply CHANGELOG.md]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/201e2b727e6f000da193c446a0d68d4914eb3b0c.

File: docs/cli-reference/dfx-canister.mdx
Comment (vincent-dfinity) [reply docs/cli-reference/dfx-canister.mdx]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/201e2b727e6f000da193c446a0d68d4914eb3b0c.

File: e2e/tests-dfx/canister_logs.bash
Comment (vincent-dfinity) [reply e2e/tests-dfx/canister_logs.bash:82]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/201e2b727e6f000da193c446a0d68d4914eb3b0c.

File: docs/cli-reference/dfx-canister.mdx
Comment (vincent-dfinity) [reply docs/cli-reference/dfx-canister.mdx]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/201e2b727e6f000da193c446a0d68d4914eb3b0c.

File: docs/cli-reference/dfx-canister.mdx
Comment (vincent-dfinity) [reply docs/cli-reference/dfx-canister.mdx]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/201e2b727e6f000da193c446a0d68d4914eb3b0c.

**Review by maksymar — APPROVED** [review]
LGTM from my side, but I suppose someone else from DFX team should take another look.

**Comment by vincent-dfinity** [comment]
> LGTM from my side, but I suppose someone else from DFX team should take another look.

Yes, Severin looked this pr. And I'm waiting for the next weekly release https://github.com/dfinity/ic/releases to update the PR, and he will take a look before merging.

File: src/dfx-core/src/config/model/dfinity.rs
Comment (sesi200) [inline src/dfx-core/src/config/model/dfinity.rs:416]:
```suggestion
    #[schemars(with = "AllowedViewers(Vec::<String>)")]
    AllowedViewers(Vec<Principal>),
```
this could save us a lot of `unwrap()`s in other places

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (sesi200) [inline src/dfx/src/lib/canister_logs/log_visibility.rs:133]: This is untrusted user input at this point, right? I don't think it's ok to just unwrap principals everywhere, but especially not if it hasn't been validated anywhere before. Can you try to find a better solution? It's probably not too hard when you use `add_log_viewer: Option<Vec<Principal>>` in the CLI opts

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (sesi200) [inline src/dfx/src/lib/canister_logs/log_visibility.rs:96]: If you make the viewers a `Set` you could save a lot of complexity when adding/removing viewers

File: src/dfx-core/src/config/model/dfinity.rs
Comment (vincent-dfinity) [reply src/dfx-core/src/config/model/dfinity.rs]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/07b2ce497e7de94506012f7e669ff78d5779543d.

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (vincent-dfinity) [reply src/dfx/src/lib/canister_logs/log_visibility.rs]: Fixed at https://github.com/dfinity/sdk/pull/3921/commits/07b2ce497e7de94506012f7e669ff78d5779543d.

File: src/dfx/src/lib/canister_logs/log_visibility.rs
Comment (vincent-dfinity) [reply src/dfx/src/lib/canister_logs/log_visibility.rs:92]: I just follow the logic of controllers. :)

**Review by sesi200 — APPROVED** [review]

**Pull request merged**
